### PR TITLE
Added search by version to releases search

### PR DESCRIPTION
**Pull request by orangejenny**
@esoergel We chatted about this a while back, and now I'm motivated to finish it because I'm planning to use `paginate_releases` to power a ajax-based version dropdown for https://github.com/dimagi/commcare-hq/pull/23214

I originally thought it'd be cleanest to do this by augmenting the ES query, but I couldn't find a way to `or` the build comment search with a version filter. `corehq.apps.queries.filtered` doesn't work because it `and`s a filter and a search, searching the version field doesn't work because it's a numeric field, and `filters.OR` seems to work with filters but not queries.

FYI @dimagi/product this adds an exact search by build version. Sample searches:
- "stuff" => All releases with "stuff" in their comment
- "35" => Version 35, if it exists, plus any builds with comments like "reverted to version 35" or "reverted to version "350". It will **not** return version 350, 135, etc.
- "stuff 35" => All releases with "stuff" and "35" in their comment. This will **not** return version 35.

code buddy @dannyroberts

### CODE REVIEW

File: corehq/apps/app_manager/views/releases.py
Comment (czue) [inline corehq/apps/app_manager/views/releases.py]: nit: this might be one of those rare situations where `try`...`else` actually makes sense assuming you only intend to catch the `ValueError` in the first line.

Comment (orangejenny) [reply corehq/apps/app_manager/views/releases.py]: good point, a85be0e9221

**Comment by orangejenny** [comment]
✅

**Comment by orangejenny** [comment]
or @mkangia

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py]: Should these two lines be within the `if` above or outside the `if build_comment`?

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py]: A comment here would be useful like
`User wants to search a specific version, entered as a build_comment`

how about renaming build_comment to search_string or query?

Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py]: Can also think about using `build_comment.is_digit()`?

Plus, i am always against extending already long methods, so if this can be extracted out, that would be great.

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py]: Can we get rid of total_apps anyway and just do len(apps) in the end to get the count?

**Review by mkangia — APPROVED** [review]
nothing blocking.

File: corehq/apps/app_manager/views/releases.py
Comment (orangejenny) [reply corehq/apps/app_manager/views/releases.py]: 8aca79122a0

File: corehq/apps/app_manager/views/releases.py
Comment (orangejenny) [reply corehq/apps/app_manager/views/releases.py]: 02a26f6d46e

Comment (orangejenny) [reply corehq/apps/app_manager/views/releases.py]: `is_digit` returns true for floats, but I only want ints.

Not going to extract the method in this PR.

File: corehq/apps/app_manager/views/releases.py
Comment (orangejenny) [reply corehq/apps/app_manager/views/releases.py]: No, apps never has more than `limit + 1` elements, but `total` needs to be the full total for the pagination "Showing X of Y" text.

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [reply corehq/apps/app_manager/views/releases.py]: I thought i checked it before suggesting. Tried again
```
In [1]: '1'.isdigit()
Out[1]: True

In [2]: '1.2'.isdigit()
Out[2]: False
```
Unless i am using a different version of python.

fine to ignore this comment.

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py]: One last query,
other apps are searched by ES and this one by couch view and hence different fields in this doc. You already pointed in the PR description. in this case,
1. The web page would break if some field is missing in this result from couch?
2. Or we would need to maintain fields in both couch view and ES result for this to keep working?

can we wait for @esoergel feedback on using ES to find the build by version?

File: corehq/apps/app_manager/views/releases.py
Comment (orangejenny) [reply corehq/apps/app_manager/views/releases.py]: Huh, I thought I tried it and `0.2` came up `True`. Not going to go back and update this one, though.

Comment (mkangia) [reply corehq/apps/app_manager/views/releases.py]: just confirmed that 0.2 also came out to be false. Just keeping note for future uses.

File: corehq/apps/app_manager/views/releases.py
Comment (orangejenny) [reply corehq/apps/app_manager/views/releases.py]: I'm not aware of us having problems with docs or fields missing from couch - couch has been reliable, just slow. What has caused problems in the past with this code is that docs have been missing from ES. So it's more likely that the search might miss a document, especially a recent one, but because search is a less-used use case and because couch-based search is so painfully slow, we can live with that.

File: corehq/apps/app_manager/views/releases.py
Comment (esoergel) [inline corehq/apps/app_manager/views/releases.py]: :+1: on the name change

**Review by esoergel — COMMENTED** [review]
You can combine queries using a `bool` query.  We have this wrapped as `add_query` (rather than `set_query`).  Just tested this out:

```python
from corehq.apps.es import queries, filters
query_string = "12"
query = es.AppES()
query = query.add_query(filters.term('version', query_string), queries.SHOULD)
query = query.add_query(queries.search_string_query(query_string, ['build_comment']), queries.SHOULD)
query.values_list("_id", "version", "build_comment")
```
I'd probably make the `build_comment` method into a standalone function implemented as above in `es/apps.py` so it's neatly wrapped, and similarly for the `version` clause.  That should do what you want, though.

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [reply corehq/apps/app_manager/views/releases.py]: ethan's suggestion might make this irrelevant but still sharing.

sorry, what i meant was the result docs from ES would have different attributes of app doc than what we would get from Couch.
Agree on missing from ES. The one and only instance i have heard of that was when one makes a build and it does not show when the page is reloaded. i think that was changed back to Couch as seen in `_get_batch` in case of no build_comment/only_show_released.

Comment (orangejenny) [reply corehq/apps/app_manager/views/releases.py]: I'm going to look at implementing Ethan's feedback, but I'm curious, why would ES and couch have different fields? If ES is behind?

**Comment by orangejenny** [comment]
@esoergel Thank you for the detailed explanation. Updated in 7e940fab44e - it's so much more concise now.

File: corehq/apps/app_manager/views/releases.py
Comment (esoergel) [reply corehq/apps/app_manager/views/releases.py]: I think Manish is referring to the transformation that sometimes happens before sending docs to ES.  This is usually the addition of a few calculated properties for certain types of queries.  For instance, we insert `owner_type` into case documents.

For apps, this is that transformation:
https://github.com/dimagi/commcare-hq/blob/master/corehq/pillows/application.py#L16-L19
That is, it applies lazy migrations, but these migrations aren't applied to docs pulled from couch unless they're wrapped.

File: corehq/apps/es/apps.py
Comment (esoergel) [inline corehq/apps/es/apps.py]: I'm not crazy about this approach - I'd expect `build_comment` to filter the results to that comment and `version` to filter results to that version.  That is, `app_es.build_comment(q).version(q)` would AND the queries together, but this `OR`s them.  What about building instead for this client code?

```python
app_es.set_query(build_comment(query), queries.SHOULD)
app_es.set_query(version(query), queries.SHOULD)
```

That would ensure that the client code knows how the queries are being joined together.

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [reply corehq/apps/app_manager/views/releases.py]: Thanks @esoergel

File: corehq/apps/es/apps.py
Comment (orangejenny) [reply corehq/apps/es/apps.py]: Like 229cab8388f? My head isn't totally wrapped around the ES logic, but I do agree that chaining functions seems like and AND operation.

I moved `build_comment` and `version` out of `AppES` to be consistent with the rest of the file, but I'd rather make them class methods. I like that the helper methods are all short names, but it's a little weird to import a method named `version` so it's now available throughout this particular file, which is likely to use the string "version" as a variable name. (There aren't actually any name conflicts, but I can imagine them being added.) What do you think?

Comment (esoergel) [reply corehq/apps/es/apps.py]: You could make a `search_comment_and_version` method on the `AppES` class if you prefer that framing (or if you think that concept is reusable).  That's an excellent point about importing - I'd use namespaces for that (I think I've done similar elsewhere)
```python
from corehq.apps.es import apps as app_es

...
        app_es.version(...)
```
That actually conflicts with the query, which is here called `app_es` (I find that name misleading, FWIW - maybe `app_query` or something), but you get the idea.

**Review by esoergel — APPROVED** [review]
I agree with your comment on name clashes, but this looks correct and I'll leave it to you to address

File: corehq/apps/es/apps.py
Comment (esoergel) [reply corehq/apps/es/apps.py]: To add to that - `version` and `build_comment` are perfectly suitable names when viewed _in the context of_ apps elasticsearch queries, and I'd prefer to bring along that namespace via the method I described above versus using a convention-based namespace like `app_es_version` and `app_es_build_comment`

**Comment by orangejenny** [comment]
Going to merge and think more about naming later.

**Pull request merged**
